### PR TITLE
Performance updates

### PR DESCRIPTION
**Pull request by louislewis2**
While working with your library, after I got my software running, I started looking for ways to improve the performance.
I then noticed you had a branch call performance. 

Before submitting performance changes, I thought it would be wise to first add some benchmark capability to ensure changes are in fact increasing performance, to monitor both the speed as well as allocations and GC pressure.

In adding this new benchmark project, I had to refactor some code to make things like assets slightly more shareable.
I did move the non src related (code that is packaged and deployed to nuget) out of the main directory and into a test directory.

While the goal was to initially just add the benchmark project with some example, I did spot one fix, would would make a dramatic and immediate change. Please see the new section in the readme file for full details.

I hope this change and addition is received well.

### CODE REVIEW

File: YoloDotNet/Data/YoloBase.cs [hunk 1]
```diff
@@ -15,6 +18,10 @@ public abstract class YoloBase : IDisposable
         protected readonly ParallelOptions _parallelOptions;
         private readonly bool _useCuda;
 
+        private int tensorBufferSize;
+        private ArrayPool<float> customSizeFloatPool;
+        protected ArrayPool<ObjectResult> customSizeObjectResultPool;
```
Comment (NickSwardh) [inline YoloDotNet/Data/YoloBase.cs:23]: Prefix private variables with an underscore _ to keep things consistent with the rest of the code

File: YoloDotNet/Data/YoloBase.cs [hunk 2]
```diff
@@ -60,6 +67,13 @@ public YoloBase(string onnxModel, bool useCuda, bool allocateGpuMemory, int gpuI
 
             if (useCuda && allocateGpuMemory)
                 _session.AllocateGpuMemory(_ortIoBinding, _runOptions);
+
+            // tensorBufferSize can be calculated once and reused for all calls, as it is based on the model properties
+            this.tensorBufferSize = OnnxModel.Input.BatchSize * OnnxModel.Input.Channels * OnnxModel.Input.Width * OnnxModel.Input.Height;
+            this.customSizeFloatPool = ArrayPool<float>.Create(maxArrayLength: this.tensorBufferSize + 1, maxArraysPerBucket: 10);
+
+
+            this.customSizeObjectResultPool = ArrayPool<ObjectResult>.Create(maxArrayLength: this.OnnxModel.Outputs[0].Channels + 1, maxArraysPerBucket: 10);
```
Comment (NickSwardh) [inline YoloDotNet/Data/YoloBase.cs:76]: `this` can be removed

File: YoloDotNet/Yolo.cs [hunk 3]
```diff
@@ -139,48 +139,56 @@ protected override ObjectResult[] ObjectDetectImage(Image image, double confiden
             var labels = OnnxModel.Labels.Length;
             var channels = OnnxModel.Outputs[0].Channels;
 
-            var boxes = new ObjectResult[channels];
+            var boxes = this.customSizeObjectResultPool.Rent(channels);
```
Comment (NickSwardh) [inline YoloDotNet/Yolo.cs:142]: `this` can be removed

File: YoloDotNet/Yolo.cs [hunk 4]
```diff
@@ -139,48 +139,56 @@ protected override ObjectResult[] ObjectDetectImage(Image image, double confiden
...
 
-            var results = boxes.Where(x => x is not null).ToArray();
+                var results = boxes.Where(x => x is not null).ToArray();
 
-            return RemoveOverlappingBoxes(results, overlapThreshold);
+                return RemoveOverlappingBoxes(results, overlapThreshold);
+            }
+            finally
+            {
+                this.customSizeObjectResultPool.Return(boxes, clearArray: true);
```
Comment (NickSwardh) [inline YoloDotNet/Yolo.cs:190]: `this` can be removed

File: YoloDotNet/Data/YoloBase.cs [hunk 5]
```diff
@@ -1,4 +1,7 @@
-﻿namespace YoloDotNet.Data
+﻿using System.Buffers;
+using YoloDotNet.Extensions;
```
Comment (NickSwardh) [inline YoloDotNet/Data/YoloBase.cs:2]: Add usings to GlobalUsings.cs

**Comment by NickSwardh** [comment]
Hi @louislewis2, thank you for contributing to YoloDotNet! Yes, benchmarking is a nice addition for testing performance, onnx-models and various code changes. The usage of an ArrayPool gave an additional boost and a much better GC performance when running multiple detections, which is great! Nice one! :)

Regarding your PR, besides my request for the minor changes I'd also like to have the `ConsoleDemo` in the the src folder with the `YoloDotNet` project for clarity, as its purpose is to demonstrate the capabilities of the core project.

Ex:
```
Solution
│
├── src
│   ├── ConsoleDemo
│   └── YoloDotNet
│
├── test
│   ├── Benchmarks
│   ├── YoloDotNet.Test.Common
│   └── Tests
```

Rock on,
Nick

**Comment by louislewis2** [comment]
@NickSwardh 

Thanks for reviewing the PR.
I have made the changes you suggested, please let me know if there is anything else I might have overlooked.

Regards

Louis

File: test/assets/Media/street640x640.jpg
Comment (NickSwardh) [inline test/assets/Media/street640x640.jpg:1]: I guess this image could be removed?

File: test/YoloDotNet.Test.Common/SharedConfig.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            ImageType.Hummingbird => Path.Combine(BASE_MEDIA, "hummingbird.jpg"),
+            ImageType.Street => Path.Combine(BASE_MEDIA, "street.jpg"),
+            ImageType.People => Path.Combine(BASE_MEDIA, "people.jpg"),
+            ImageType.Crosswalk => Path.Combine(BASE_MEDIA, "crosswalk.jpg"),
+            ImageType.Island => Path.Combine(BASE_MEDIA, "island.jpg"),
+            _ => throw new ArgumentException("Unknown modeltype.")
+        };
+
+        public static string GetTestImage(string imageName) => Path.Combine(BASE_MEDIA, imageName);
```
Comment (NickSwardh) [inline test/YoloDotNet.Test.Common/SharedConfig.cs:35]: The `Assets` folder and its contents is currently not visible to the user in the Solution Explorer. The `Assets` folder could be moved into the `YoloDotNet.Test.Common` for a more centralized configuration. To get access to the images and models, `SharedConfig.cs` could then be updated to something like this:
```csharp
private const string _projectName = nameof(YoloDotNet);

private static readonly string _dir = AppDomain.CurrentDomain.BaseDirectory;
private static readonly string _baseDir = _dir[0..(_dir.IndexOf(_projectName) + _projectName.Length)];

private static readonly string _assetsFolder = Path.Combine(_baseDir, @"test\YoloDotNet.Test.Common\Assets");
private static readonly string _baseModels = Path.Combine(_assetsFolder, "Models");
private static readonly string _baseMedia = Path.Combine(_assetsFolder, "Media");

public static string GetTestModel(ModelType modelType) => modelType switch
{
    ModelType.Classification => Path.Combine(_baseModels, "yolov8s-cls.onnx"),
    ModelType.ObjectDetection => Path.Combine(_baseModels, "yolov8s.onnx"),
    ModelType.ObbDetection => Path.Combine(_baseModels, "yolov8s-obb.onnx"),
    ModelType.Segmentation => Path.Combine(_baseModels, "yolov8s-seg.onnx"),
    ModelType.PoseEstimation => Path.Combine(_baseModels, "yolov8s-pose.onnx"),
    _ => throw new ArgumentException("Unknown modeltype.")
};

public static string GetTestImage(ImageType imageType) => imageType switch
{
    ImageType.Hummingbird => Path.Combine(_baseMedia, "hummingbird.jpg"),
    ImageType.Street => Path.Combine(_baseMedia, "street.jpg"),
    ImageType.People => Path.Combine(_baseMedia, "people.jpg"),
    ImageType.Crosswalk => Path.Combine(_baseMedia, "crosswalk.jpg"),
    ImageType.Island => Path.Combine(_baseMedia, "island.jpg"),
    _ => throw new ArgumentException("Unknown test image.")
};

public static string GetTestImage(string imageName) => Path.Combine(_baseMedia, imageName);
```

**Review by NickSwardh — COMMENTED** [review]
Thank you @louislewis2 for the changes. Found a few other things to have a look at. Also, not a big dealt but just to stay consistens with the rest of the code, if you could update the benchmarks and the private variables with a prefixed unserscore and remove `this` would be great. Rock on dude! :)

File: test/assets/Media/street640x640.jpg
Comment (louislewis2) [reply test/assets/Media/street640x640.jpg:1]: It is being used by a benchmark, which would be extended later on.
The benchmarks at the moment, perform a single operation as there is enough to fix based on the results.
As the code becomes better in terms of the benchmark, one could perform the operation multiple times, which would surface more detail. 

While the resize benchmark for a single run, might not yield much in terms of time spent or allocations, doing that a few times surfaces the true results. In my example, I am processing around 60 frames per second here the resize overhead plays a large role.

Hope that makes sense as to why it is still there.

**Pull request merged**